### PR TITLE
Submit transactions to follower

### PR DESCRIPTION
**Pull request by kc1212**
We implement this functionality using a polling leader. Every follower
(including the leader) stores new transactions in a temporary buffer.
This buffer is emptied every time a polling request arrives and the
content of the buffer is sent to the leader. Followers are responsible
for checking whether they are sending the real leader. Otherwise a
malicious polling request may cause the follower to empty its buffer
prematurely.

Closes #1329

### CODE REVIEW

File: omniledger/service/service.go [hunk 1]
```diff
@@ -28,6 +27,9 @@ const invokeEvolve darc.Action = darc.Action("invoke:evolve")
 
 // OmniledgerID can be used to refer to this service
 var OmniledgerID onet.ServiceID
+
+const collectTxProtocol = "CollectTxProtocol"
```
Comment (ineiti) [inline omniledger/service/service.go]: that should be in `collect_tx.go`

File: omniledger/service/service.go [hunk 2]
```diff
@@ -396,6 +400,18 @@ func (s *Service) updateCollection(msg network.Message) {
 		log.Error("Couldn't recreate state changes:", err.Error())
 		return
 	}
+
+	// Check whether we are storing a genesis config, if we are, also set
+	// the leaderMap.
+	if scs.isGenesisConfig() {
```
Comment (ineiti) [inline omniledger/service/service.go]: Why is that here and not in the `Config`-contract?

File: omniledger/service/leader_map.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
+package service
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/dedis/onet/network"
+)
+
+type leaderMap struct {
```
Comment (ineiti) [inline omniledger/service/leader_map.go]: I don't understand why you need this `leaderMap` - can't you take the roster of the skipchain and look at the first entry?

File: omniledger/service/service.go (hunk 1 above)
Comment (kc1212) [reply omniledger/service/service.go]: I put it here because `CollectTx` is not a "full" protocol, it needs to be registered with a `getTxs` callback. This is similar to how byzcoinx does it.

File: omniledger/service/service.go (hunk 2 above)
Comment (kc1212) [reply omniledger/service/service.go]: probably will be removed if `leaderMap` is removed

**Comment by kc1212** [comment]
@ineiti removed `leaderMap` and `isGenesisConfig`

File: omniledger/service/collect_tx.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+		return errors.New("did not receive request")
+	}
+
+	// send the result of the callback to the root
+	resp := &CollectTxResponse{
+		Txs: p.getTxs(req.ServerIdentity, req.SkipchainID),
+	}
+	if p.IsRoot() {
+		if err := p.SendTo(p.TreeNode(), resp); err != nil {
+			return err
```
Comment (jeffallen) [inline omniledger/service/collect_tx.go]: If either of these error cases happen, then the transactions which were taken get lost, because they are not re-put back.

File: omniledger/service/service.go [hunk 5]
```diff
@@ -473,53 +476,66 @@ func (s *Service) loadLatestDarc(sid skipchain.SkipBlockID, dID darc.ID) (*darc.
...
+				if err != nil {
+					panic("DB is in bad state and cannot find skipchain anymore: " + err.Error())
+				}
+				// We assume the caller of this function is the
+				// current leader. So we generate a tree with
+				// this node as the root.
+				tree := sb.Roster.GenerateNaryTreeWithRoot(len(sb.Roster.List), s.ServerIdentity())
+				proto, err := s.CreateProtocol(collectTxProtocol, tree)
+				if err != nil {
+					panic("cannot create protocol: " + err.Error())
```
Comment (jeffallen) [inline omniledger/service/service.go]: If someone, somehow, finds a way to make this happen, then the entire server will crash. There's got to be a better reaction than panic here., like maybe log.Error() and close(closeSignal).

File: omniledger/service/service.go [hunk 6]
```diff
@@ -622,6 +638,37 @@ clientTransactions:
...
+	if err != nil {
+		return nil, err
+	}
+	if sb.Roster == nil || len(sb.Roster.List) < 1 {
+		return nil, errors.New("roster is empty")
+	}
+	return sb.Roster.List[0], nil
+}
+
+func (s *Service) getTxs(leader *network.ServerIdentity, scID skipchain.SkipBlockID) ClientTransactions {
```
Comment (jeffallen) [inline omniledger/service/service.go]: Why does getTxs even care who the current leader is?

File: omniledger/service/service_test.go [hunk 7]
```diff
@@ -65,6 +65,14 @@ func padDarc(key []byte) []byte {
...
 
 func TestService_AddTransaction(t *testing.T) {
+	testAddTransaction(t, 0)
+}
+
+func TestService_AddTransactionToFollower(t *testing.T) {
+	testAddTransaction(t, 1)
+}
+
+func testAddTransaction(t *testing.T, sendToIdx int) {
```
Comment (jeffallen) [inline omniledger/service/service_test.go]: This is really nice and clear, I like.

File: omniledger/service/service.go (hunk 6 above)
Comment (ineiti) [reply omniledger/service/service.go]: So it can verify if it is the _real_ leader?

File: omniledger/service/collect_tx.go (hunk 4 above)
Comment (kc1212) [reply omniledger/service/collect_tx.go]: This is the intended behaviour when I was implementing it. We cannot guarantee every transaction that gets submitted to OL will eventually be included in the skipchain. It'll be even harder to have this guarantee when the leader starts to change (I don't know how to do it yet). But in this particular case, it should be possible to put the transactions back if there is an error.

File: omniledger/service/service.go (hunk 5 above)
Comment (kc1212) [reply omniledger/service/service.go]: The reason I decided to panic is that there is simply nothing we can do to recover without putting the server in a strange state. If you stop the go-routine than that would result in a non-polling leader. Then somebody has to go and fix the problem and restart the server.

File: skipchain/skipchain.go [hunk 8]
```diff
@@ -821,11 +821,16 @@ func (s *Service) SetBFTTimeout(t time.Duration) {
...
 }
 
-// SetPropTimeout is used to set the propagation timeout
+// SetPropTimeout is used to set the propagation timeout.
 func (s *Service) SetPropTimeout(t time.Duration) {
 	s.propTimeout = t
 }
 
+// GetPropTimeout retrieves the current propagation timeout.
+func (s *Service) GetPropTimeout() time.Duration {
```
[the remark this answers is not in the document]
Comment (kc1212) [reply skipchain/skipchain.go]: I think `SetPropTimeout` is only meant to be used in tests, only in skipchain tests IIRC, so I didn't put a lock in `GetPropTimeout`.

File: omniledger/service/collect_tx.go (hunk 4 above)
Comment (kc1212) [reply omniledger/service/collect_tx.go]: Created an issue to track this problem https://github.com/dedis/cothority/issues/1337

File: omniledger/service/service.go (hunk 5 above)
Comment (jeffallen) [reply omniledger/service/service.go]: ok, good reasoning. Can you put it in the code, so that later when we say, "damn it, why is this panic here?" we don't forget?

**Comment by kc1212** [comment]
Moving back to WIP because there seems to be a data race in collections...

**Comment by kc1212** [comment]
Can't produce the race condition anymore... moving back to R4M, looks like there's already an issue tracking it https://github.com/dedis/cothority/issues/1317

File: omniledger/service/service.go (hunk 5 above)
Comment (kc1212) [reply omniledger/service/service.go]: sure, I updated it
